File: js/entities.js
class Player {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = 30;
        this.height = 30;
        this.speed = 4;
        this.color = '#0ff';
        this.cooldownAir = 0;
        this.cooldownGround = 0;
        this.reticleDistance = 120;
    }

    update(keys, canvasWidth, canvasHeight) {
        // Movement
        if (keys['ArrowUp'] || keys['ArrowUp']) this.y -= this.speed;
        if (keys['ArrowDown'] || keys['ArrowDown']) this.y += this.speed;
        if (keys['ArrowLeft'] || keys['ArrowLeft']) this.x -= this.speed;
        if (keys['ArrowRight'] || keys['ArrowRight']) this.x += this.speed;

        // Boundaries
        this.x = Math.max(this.width/2, Math.min(canvasWidth - this.width/2, this.x));
        this.y = Math.max(this.height/2, Math.min(canvasHeight - this.height/2, this.y));

        // Cooldowns
        if (this.cooldownAir > 0) this.cooldownAir--;
        if (this.cooldownGround > 0) this.cooldownGround--;
    }

    draw(ctx) {
        // Draw Player Ship (Geometric shape: triangle/polygon)
        ctx.fillStyle = this.color;
        ctx.beginPath();
        ctx.moveTo(this.x, this.y - this.height / 2); // Nose
        ctx.lineTo(this.x + this.width / 2, this.y + this.height / 2); // Right wing
        ctx.lineTo(this.x, this.y + this.height / 4); // Engine indent
        ctx.lineTo(this.x - this.width / 2, this.y + this.height / 2); // Left wing
        ctx.closePath();
        ctx.fill();

        // Draw Reticle
        this.drawReticle(ctx);
    }

    drawReticle(ctx) {
        const reticleX = this.x;
        const reticleY = this.y - this.reticleDistance;
        
        ctx.strokeStyle = 'rgba(255, 0, 0, 0.7)';
        ctx.lineWidth = 2;
        ctx.beginPath();
        // Crosshair
        ctx.moveTo(reticleX - 10, reticleY);
        ctx.lineTo(reticleX + 10, reticleY);
        ctx.moveTo(reticleX, reticleY - 10);
        ctx.lineTo(reticleX, reticleY + 10);
        
        // Box
        ctx.strokeRect(reticleX - 8, reticleY - 8, 16, 16);
        ctx.stroke();
    }
}

class Bullet {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = 4;
        this.height = 12;
        this.speed = 10;
        this.color = '#ff0';
        this.markedForDeletion = false;
    }

    update() {
        this.y -= this.speed;
        if (this.y < 0) this.markedForDeletion = true;
    }

    draw(ctx) {
        ctx.fillStyle = this.color;
        ctx.fillRect(this.x - this.width / 2, this.y - this.height / 2, this.width, this.height);
    }
}

class Bomb {
    constructor(startX, startY, targetX, targetY) {
        this.x = startX;
        this.y = startY;
        this.targetX = targetX;
        this.targetY = targetY;
        
        this.progress = 0; // 0 to 1
        this.speed = 0.03;
        this.size = 10;
        this.markedForDeletion = false;
        this.exploded = false;
        this.explosionTimer = 0;
    }

    update() {
        if (this.exploded) {
            this.explosionTimer++;
            if (this.explosionTimer > 15) {
                this.markedForDeletion = true;
            }
            return;
        }

        this.progress += this.speed;
        
        // Lerp position
        this.x = this.x + (this.targetX - this.x) * 0.1;
        this.y = this.y + (this.targetY - this.y) * this.progress;

        if (this.progress >= 1) {
            this.progress = 1;
            if (!this.exploded) {
                this.exploded = true;
                audioManager.playExplosionGround();
            }
            this.y = this.targetY; // snap to target
            this.x = this.targetX;
        }
    }

    draw(ctx) {
        if (this.exploded) {
            // Draw explosion
            ctx.fillStyle = '#f80';
            ctx.beginPath();
            ctx.arc(this.x, this.y, 20 + this.explosionTimer, 0, Math.PI * 2);
            ctx.fill();
            
            ctx.fillStyle = '#ff0';
            ctx.beginPath();
            ctx.arc(this.x, this.y, 10 + this.explosionTimer * 0.5, 0, Math.PI * 2);
            ctx.fill();
        } else {
            // Draw falling bomb
            const currentSize = this.size * (1 - this.progress * 0.5); // Gets smaller as it falls
            ctx.fillStyle = '#f0f';
            ctx.beginPath();
            ctx.arc(this.x, this.y, currentSize, 0, Math.PI * 2);
            ctx.fill();
        }
    }
}

class Enemy {
    constructor(x, y, type) {
        this.x = x;
        this.y = y;
        this.type = type; // 'air' or 'ground'
        this.markedForDeletion = false;
        
        if (type === 'air') {
            this.width = 24;
            this.height = 24;
            this.color = '#f00';
            this.speed = 2;
            this.angle = 0;
            this.movementType = Math.floor(Math.random() * 3);
        } else {
            // Ground target
            this.width = 32;
            this.height = 32;
            this.color = '#a0a';
            this.speed = 1; // Scrolls with background
        }
    }

    update() {
        if (this.type === 'air') {
            switch(this.movementType) {
                case 0: // Straight down
                    this.y += this.speed;
                    break;
                case 1: // Sine wave
                    this.y += this.speed * 0.8;
                    this.x += Math.sin(this.y * 0.05) * 3;
                    break;
                case 2: // Diagonal
                    this.y += this.speed;
                    this.x += this.speed * (this.x > 240 ? -0.5 : 0.5);
                    break;
            }
        } else {
            // Ground targets just scroll down
            this.y += this.speed;
        }

        if (this.y > 700) {
            this.markedForDeletion = true;
        }
    }

    draw(ctx) {
        ctx.fillStyle = this.color;
        if (this.type === 'air') {
            // Draw UFO shape
            ctx.beginPath();
            ctx.ellipse(this.x, this.y, this.width/2, this.height/4, 0, 0, Math.PI * 2);
            ctx.fill();
            ctx.fillStyle = '#0f0'; // cockpit
            ctx.beginPath();
            ctx.arc(this.x, this.y - 4, this.width/4, 0, Math.PI * 2);
            ctx.fill();
        } else {
            // Draw ground pyramid/base
            ctx.fillRect(this.x - this.width/2, this.y - this.height/2, this.width, this.height);
            ctx.fillStyle = '#f00';
            ctx.beginPath();
            ctx.arc(this.x, this.y, 8, 0, Math.PI * 2);
            ctx.fill();
        }
    }
}

class Background {
    constructor(width, height) {
        this.width = width;
        this.height = height;
        this.y = 0;
        this.speed = 1;
        // Generate some basic terrain elements
        this.elements = [];
        for (let i = 0; i < 20; i++) {
            this.elements.push({
                x: Math.random() * width,
                y: Math.random() * height * 2 - height, // Pre-fill above and current screen
                size: Math.random() * 40 + 20,
                color: Math.random() > 0.5 ? '#030' : '#040' // Forest greens
            });
        }
        
        // Add a "river"
        this.riverPoints = [];
        for (let i = 0; i < 20; i++) {
            this.riverPoints.push(Math.sin(i * 0.5) * 40 + width / 2);
        }
    }

    update() {
        this.y += this.speed;
        if (this.y > this.height) {
            this.y = 0;
        }
        
        // Recycle terrain elements
        this.elements.forEach(el => {
            el.y += this.speed;
            if (el.y > this.height) {
                el.y = -el.size;
                el.x = Math.random() * this.width;
            }
        });
    }

    draw(ctx) {
        // Base ground color
        ctx.fillStyle = '#002200';
        ctx.fillRect(0, 0, this.width, this.height);

        // Draw terrain elements (trees/forests)
        this.elements.forEach(el => {
            ctx.fillStyle = el.color;
            ctx.fillRect(el.x, el.y, el.size, el.size);
            // Draw wrap-around for smooth scrolling
            ctx.fillRect(el.x, el.y - this.height, el.size, el.size);
        });

        // Draw River
        ctx.strokeStyle = '#0000aa';
        ctx.lineWidth = 30;
        ctx.lineJoin = 'round';
        
        ctx.beginPath();
        for(let i=0; i < 20; i++) {
            let ry = (i * (this.height/10)) + this.y % (this.height/10);
            if (i === 0) ctx.moveTo(this.riverPoints[i], ry);
            else ctx.lineTo(this.riverPoints[i], ry);
        }
        ctx.stroke();
    }
}

File: entities.py
import pygame
import math
import random
from audio import audio_manager

class Player:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 30
        self.height = 30
        self.speed = 4
        self.color = (0, 255, 255)
        self.cooldown_air = 0
        self.cooldown_ground = 0
        self.reticle_distance = 120
        # Charge Hadouho states
        self.charge_timer = 0
        self.charge_max = 60
        self.was_z_pressed = False
        self.shoot_normal = False
        self.shoot_wave = False
        self.played_complete = False
        self.shield_count = 0
        self.weapon_level = 1

    def update(self, keys, canvas_width, canvas_height):
        if keys.get(pygame.K_UP): self.y -= self.speed
        if keys.get(pygame.K_DOWN): self.y += self.speed
        if keys.get(pygame.K_LEFT): self.x -= self.speed
        if keys.get(pygame.K_RIGHT): self.x += self.speed

        self.x = max(self.width/2, min(canvas_width - self.width/2, self.x))
        self.y = max(self.height/2, min(canvas_height - self.height/2, self.y))

        if self.cooldown_air > 0: self.cooldown_air -= 1
        if self.cooldown_ground > 0: self.cooldown_ground -= 1

        # Hadouho charging input logic
        z_pressed = keys.get(pygame.K_z, False) or keys.get(pygame.K_c, False)
        
        if z_pressed:
            if not self.was_z_pressed:
                if self.cooldown_air <= 0:
                    self.shoot_normal = True
            
            self.charge_timer = min(self.charge_max, self.charge_timer + 1)
            
            if self.charge_timer == 1:
                audio_manager.play_charge()
            elif self.charge_timer == self.charge_max:
                if not getattr(self, 'played_complete', False):
                    audio_manager.play('charge_complete')
                    self.played_complete = True
        else:
            if self.was_z_pressed:
                audio_manager.stop_charge()
                if self.charge_timer >= self.charge_max:
                    self.shoot_wave = True
                self.charge_timer = 0
                self.played_complete = False
                
        self.was_z_pressed = z_pressed

    def draw(self, surface):
        # Draw charging aura if charging
        if self.charge_timer > 0:
            charge_ratio = self.charge_timer / self.charge_max
            pulse = math.sin(pygame.time.get_ticks() * 0.02) * 5
            radius = int(25 * charge_ratio + 5 + pulse)
            
            if charge_ratio >= 1.0:
                color = (255, 255, 255) if (pygame.time.get_ticks() // 50) % 2 == 0 else (0, 255, 255)
                pygame.draw.circle(surface, color, (int(self.x), int(self.y)), radius, 2)
                pygame.draw.circle(surface, (0, 170, 255), (int(self.x), int(self.y)), radius - 4, 1)
            else:
                color = (0, 100, 255)
                pygame.draw.circle(surface, color, (int(self.x), int(self.y)), radius, 1)

        points = [
            (self.x, self.y - self.height / 2),
            (self.x + self.width / 2, self.y + self.height / 2),
            (self.x, self.y + self.height / 4),
            (self.x - self.width / 2, self.y + self.height / 2)
        ]
        pygame.draw.polygon(surface, self.color, points)

        if getattr(self, 'shield_count', 0) > 0:
            pygame.draw.circle(surface, (0, 255, 255), (int(self.x), int(self.y)), int(self.width / 2 + 8), 2)
            for i in range(self.shield_count):
                angle = i * (2 * math.pi / self.shield_count) + pygame.time.get_ticks() * 0.005
                hx = self.x + math.cos(angle) * (self.width / 2 + 8)
                hy = self.y + math.sin(angle) * (self.width / 2 + 8)
                pygame.draw.circle(surface, (255, 255, 255), (int(hx), int(hy)), 3)

        self.draw_reticle(surface)

    def draw_reticle(self, surface):
        reticle_x = self.x
        reticle_y = self.y - self.reticle_distance
        red = (255, 0, 0)
        
        pygame.draw.line(surface, red, (reticle_x - 10, reticle_y), (reticle_x + 10, reticle_y), 2)
        pygame.draw.line(surface, red, (reticle_x, reticle_y - 10), (reticle_x, reticle_y + 10), 2)
        pygame.draw.rect(surface, red, (reticle_x - 8, reticle_y - 8, 16, 16), 2)

class Bullet:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 4
        self.height = 12
        self.speed = 10
        self.color = (255, 255, 0)
        self.marked_for_deletion = False

    def update(self):
        self.y -= self.speed
        if self.y < 0: self.marked_for_deletion = True

    def draw(self, surface):
        pygame.draw.rect(surface, self.color, (self.x - self.width / 2, self.y - self.height / 2, self.width, self.height))

class EnemyBullet:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 6
        self.height = 6
        self.speed = 5
        self.color = (255, 100, 100)
        self.marked_for_deletion = False

    def update(self):
        self.y += self.speed
        if self.y > 700: self.marked_for_deletion = True

    def draw(self, surface):
        pygame.draw.circle(surface, self.color, (int(self.x), int(self.y)), int(self.width / 2))

class WaveCannon:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 32
        self.height = 32
        self.speed = 12
        self.color = (0, 255, 255)
        self.marked_for_deletion = False

    def update(self):
        self.y -= self.speed
        if self.y < -50: self.marked_for_deletion = True

    def draw(self, surface):
        pulse = int(math.sin(pygame.time.get_ticks() * 0.05) * 4)
        pygame.draw.circle(surface, (255, 255, 255), (int(self.x), int(self.y)), 12 + pulse)
        pygame.draw.circle(surface, (0, 170, 255), (int(self.x), int(self.y)), 22 + pulse, 3)
        pygame.draw.ellipse(surface, (0, 255, 255), (self.x - 30, self.y - 8, 60, 16), 2)

class Bomb:
    def __init__(self, start_x, start_y, target_x, target_y):
        self.x = start_x
        self.y = start_y
        self.target_x = target_x
        self.target_y = target_y
        
        self.progress = 0.0
        self.speed = 0.03
        self.size = 10
        self.marked_for_deletion = False
        self.exploded = False
        self.explosion_timer = 0

    def update(self):
        if self.exploded:
            self.explosion_timer += 1
            if self.explosion_timer > 15:
                self.marked_for_deletion = True
            return

        self.progress += self.speed
        
        self.x = self.x + (self.target_x - self.x) * 0.1
        self.y = self.y + (self.target_y - self.y) * self.progress

        if self.progress >= 1.0:
            self.progress = 1.0
            if not self.exploded:
                self.exploded = True
                audio_manager.play('explosion_ground')
            self.y = self.target_y
            self.x = self.target_x

    def draw(self, surface):
        if self.exploded:
            pygame.draw.circle(surface, (255, 136, 0), (int(self.x), int(self.y)), 20 + self.explosion_timer)
            pygame.draw.circle(surface, (255, 255, 0), (int(self.x), int(self.y)), int(10 + self.explosion_timer * 0.5))
        else:
            current_size = max(1, int(self.size * (1 - self.progress * 0.5)))
            pygame.draw.circle(surface, (255, 0, 255), (int(self.x), int(self.y)), current_size)

class Enemy:
    def __init__(self, x, y, type_):
        self.x = x
        self.y = y
        self.type = type_
        self.marked_for_deletion = False
        self.shoot_timer = random.randint(30, 120)
        self.shoot_now = False
        
        if type_ == 'air':
            self.width = 24
            self.height = 24
            self.color = (255, 0, 0)
            self.speed = 2
            self.movement_type = random.randint(0, 2)
        else:
            self.width = 32
            self.height = 32
            self.color = (170, 0, 170)
            self.speed = 1

    def update(self):
        if self.type == 'air':
            self.shoot_timer -= 1
            if self.shoot_timer <= 0:
                self.shoot_now = True
                self.shoot_timer = random.randint(60, 180)

            if self.movement_type == 0:
                self.y += self.speed
            elif self.movement_type == 1:
                self.y += self.speed * 0.8
                self.x += math.sin(self.y * 0.05) * 3
            elif self.movement_type == 2:
                self.y += self.speed
                self.x += self.speed * (-0.5 if self.x > 240 else 0.5)
        else:
            self.y += self.speed

        if self.y > 700:
            self.marked_for_deletion = True

    def draw(self, surface):
        if self.type == 'air':
            rect = (self.x - self.width/2, self.y - self.height/4, self.width, self.height/2)
            pygame.draw.ellipse(surface, self.color, rect)
            pygame.draw.circle(surface, (0, 255, 0), (int(self.x), int(self.y - 4)), int(self.width/4))
        else:
            pygame.draw.rect(surface, self.color, (self.x - self.width/2, self.y - self.height/2, self.width, self.height))
            pygame.draw.circle(surface, (255, 0, 0), (int(self.x), int(self.y)), 8)

class Background:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.y = 0
        self.speed = 1
        
        self.elements = []
        for _ in range(20):
            self.elements.append({
                'x': random.uniform(0, width),
                'y': random.uniform(-height, height),
                'size': random.uniform(20, 60),
                'color': (0, 51, 0) if random.random() > 0.5 else (0, 68, 0)
            })
            
        self.river_points = [math.sin(i * 0.5) * 40 + width / 2 for i in range(20)]

    def update(self):
        self.y += self.speed
        if self.y > self.height:
            self.y = 0
            
        for el in self.elements:
            el['y'] += self.speed
            if el['y'] > self.height:
                el['y'] = -el['size']
                el['x'] = random.uniform(0, self.width)

    def draw(self, surface):
        surface.fill((0, 34, 0))

        for el in self.elements:
            color = el['color']
            x, y, size = el['x'], el['y'], el['size']
            pygame.draw.rect(surface, color, (x, y, size, size))
            pygame.draw.rect(surface, color, (x, y - self.height, size, size))
            
        points = []
        for i in range(20):
            ry = (i * (self.height / 10)) + self.y % (self.height / 10)
            points.append((self.river_points[i], ry))
            
        if len(points) > 1:
            pygame.draw.lines(surface, (0, 0, 170), False, points, 30)

class ShieldCapsule:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 16
        self.height = 16
        self.speed = 1.5
        self.marked_for_deletion = False

    def update(self):
        self.y += self.speed
        if self.y > 700: self.marked_for_deletion = True

    def draw(self, surface):
        pygame.draw.ellipse(surface, (0, 200, 255), (self.x - self.width / 2, self.y - self.height / 2, self.width, self.height))
        pygame.draw.circle(surface, (255, 255, 255), (int(self.x - 2), int(self.y - 2)), 3)

class WeaponCapsule:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 16
        self.height = 16
        self.speed = 1.5
        self.marked_for_deletion = False

    def update(self):
        self.y += self.speed
        if self.y > 700: self.marked_for_deletion = True

    def draw(self, surface):
        pygame.draw.ellipse(surface, (255, 50, 50), (self.x - self.width / 2, self.y - self.height / 2, self.width, self.height))
        pygame.draw.circle(surface, (255, 255, 255), (int(self.x - 2), int(self.y - 2)), 3)

